### DISCUSSION TITLE
Python code questions

### DISCUSSION THREAD

**Discussion by RunSongham** [pyocd/pyOCD#1679] [Q&A]
Hi.

Developing a debug tool using gui as python.
Now you need to connect the j-link to perform the MCU flash write.
As I looked for information, I found out pyocd, and as I read the document, I don't know what I want in detail, so I ask.
1. I want to check the debugger list in Python code like "pyocd list", which one should I refer to?
2. I want to check pyocd mcu pack list in python code. Which one should I refer to?
![image](https://github.com/pyocd/pyOCD/assets/16534811/b47e02be-4cdd-4847-a632-604cb0b2043b)

**Comment by elfmimi** [comment]
Here you go.

1. pyocd list ⇒ [list_cmd.py](https://github.com/pyocd/pyOCD/blob/main/pyocd/subcommands/list_cmd.py) , [helper.py](https://github.com/pyocd/pyOCD/blob/5ab5c327987c4c3ba5a37928cac25263d4db7586/pyocd/core/helpers.py#L268)

```python
from pyocd.core.helpers import  ConnectHelper
import prettytable

probes = ConnectHelper.get_all_connected_probes(blocking=False)

pt = prettytable.PrettyTable(["#", "Probe/Board", "Unique ID"])
pt.align = 'l'
pt.header = True
pt.border = True
pt.hrules = prettytable.HEADER
pt.vrules = prettytable.NONE

for index, probe in enumerate(probes):
    pt.add_row([
        str(index),
        probe.product_name,
        probe.unique_id,
    ])

print(pt)
```

2. pyocd pack show ⇒ [pack_cmd.py](https://github.com/pyocd/pyOCD/blob/5ab5c327987c4c3ba5a37928cac25263d4db7586/pyocd/subcommands/pack_cmd.py#L133)
```python
from pyocd.target.pack import pack_target
import prettytable

packs = pack_target.ManagedPacks.get_installed_packs() 

pt = prettytable.PrettyTable(["Pack", "Version", "Pack File"])
pt.align = 'l'
pt.header = True
pt.border = True
pt.hrules = prettytable.HEADER
pt.vrules = prettytable.NONE

for ref in packs:
    pt.add_row([
        f"{ref.vendor}.{ref.pack}",
        ref.version,
        ref.get_pack_name(),
    ])

print(pt)
```
